[클린코드 4기 유수아] 계산기 구현

aken-you opened the pull request:
안녕하세요. 리뷰어님!

class, cypress를 처음 접해서 공식 문서 보면서 할 수 있는 한 구현해봤습니다. 미션하면서 js 배운지 얼마 되지 않아 모듈, 함수 분리하고 변수명 짓는데 어려움이 있었습니다. 더 좋은 방법이 있다면 피드백 부탁드립니다!

### 질문

1. calculator.js 파일 80번째 줄에 `else`를 썼는데 안 쓰고 할 수 있는 좋은 코드는 없을까요? `return`을 써서 `else`를 없애보려 했는데 두 경우 다 `this.renderTotal()` 실행해야 해서 코드가 더 늘어났습니다. 좋은 방법이 있다면 알려주세요!
2. class method 안에서 addEventListener의 callback 함수를 사용하는 방법 중 좋은 방법은 무엇인가요? [이 사이트](https://stackoverflow.com/questions/43727516/how-adding-event-handler-inside-a-class-with-a-class-method-as-the-callback?noredirect=1&lq=1)를 통해 저는 화살표 함수로 함수를 사용했습니다.
3. cypress test에서 숫자, 연산자를 스스로 정하다 보니 다른 예외 경우를 확인하지 못할 수 있다고 생각했습니다. 보통 test에서 숫자나 연산자를 정하는 함수를 만드는지 궁금합니다.

aken-you committed:
chore: 패키지 설정

- eslint, prettier 설정
- cypress 설치

aken-you committed:
test: 사칙 연산 테스트

aken-you committed:
feat: 사칙연산 기능 구현

aken-you committed:
test: 초기화 기능 테스트

aken-you committed:
feat: 초기화 기능 구현

aken-you committed:
test: 테스트 추가

- 이전 결과값을 이용해 계속 계산할 수 있는지 확인하는 테스트
- 3자리 숫자를 넘어가면 경고창이 뜨는지 확인하는 테스트

aken-you committed:
feat: 숫자를 최대 3자리 수까지 입력 제한하는 기능 구현

aken-you committed:
refactor: return 이동, clickDigit함수에서 메소드 분리

aken-you committed:
test: 사칙 연산자를 누르고 = 버튼 누르면 경고창 뜨는지 확인하는 테스트

aken-you committed:
feat: 사칙 연산자(+, -, X, /)를 누른 후 = 버튼 클릭 제한하는 기능 구현

aken-you committed:
test: 덧셈 테스트 추가

aken-you committed:
style: 함수명 변경

yujo11 commented on .eslintrc.js:
```diff
@@ -0,0 +1,18 @@
+  },
+  rules: {
+    'import/extensions': 'off',
+    'max-depth': ['error', 2],
```
오 `max-detph` rule`까지 설정하셨군요! 멋집니다!! 👍 👍 👍

yujo11 commented on README.md:
```diff
@@ -26,13 +26,13 @@
-- [ ] AC(All Clear)버튼을 누르면 0으로 초기화 한다.
-- [ ] 숫자는 한번에 최대 3자리 수까지 입력 가능하다.
-- [ ] 계산 결과를 표현할 때 소수점 이하는 버림한다.
+- [x] 2개의 숫자에 대해 덧셈이 가능하다.
+- [x] 2개의 숫자에 대해 뺄셈이 가능하다.
+- [x] 2개의 숫자에 대해 곱셈이 가능하다.
+- [x] 2개의 숫자에 대해 나눗셈이 가능하다.
+- [x] 계산 결과를 표현할 때 소수점 이하는 버림한다.
+- [x] AC(All Clear)버튼을 누르면 0으로 초기화 한다.
+- [x] 숫자는 한번에 최대 3자리 수까지 입력 가능하다.
```
👍👍👍

yujo11 commented on cypress/support/commands.js:
```diff
@@ -0,0 +1,37 @@
+//
+// -- This will overwrite an existing command --
+// Cypress.Commands.overwrite('visit', (originalFn, url, options) => { ... })
+Cypress.Commands.add('clickDigit', (numbers) => {
+  numbers.forEach((number) => cy.get(`button[data-digit="${number}"]`).click());
+});
+Cypress.Commands.add('clickOperation', (operation) => {
+  cy.get(`button[data-operation="${operation}"]`).click();
+});
+Cypress.Commands.add('clickModifier', () => {
+  cy.get(`button[data-modifier="AC"]`).click();
+});
+Cypress.Commands.add('getTotal', (total) => {
+  cy.get('#total').should('have.text', `${total}`);
+});
```
오 Command를 추가해서 사용하셨군요! 👍 👍 👍

yujo11 commented on cypress/e2e/calculator.cy.js:
```diff
@@ -0,0 +1,222 @@
+
+  describe('2개의 숫자에 대해 덧셈이 가능하다.', () => {
+    it('22 + 222 = 244', () => {
+      cy.clickDigit([2, 2]);
```
`clickDigit` 커맨드가 유용해 보이네요! 👍 👍 👍

yujo11 commented on cypress/e2e/calculator.cy.js:
```diff
@@ -0,0 +1,222 @@
+      cy.clickDigit([0]);
+      cy.clickOperation(OPERATION.EQUAL);
+      cy.getTotal(400);
+    });
```
`숫자 클릭 -> 연산자 클릭 -> 결과 확인` 이라는 일련의 과정이 반복되고 있는데요.

이 부분도 추상화해서 Command로 만들어보면 어떨까요?

aken-you replied:
다른 중복되는 코드도 Command로 만들어 보겠습니다!

yujo11 commented on cypress/e2e/calculator.cy.js:
```diff
@@ -0,0 +1,222 @@
+      cy.clickDigit([0]);
+      cy.clickOperation(OPERATION.EQUAL);
+      cy.getTotal(Infinity);
+    });
```
예외사항까지 꼼꼼하게 테스트해보셨군요! 👍 👍 👍

yujo11 commented on cypress/e2e/calculator.cy.js:
```diff
@@ -0,0 +1,222 @@
+    });
+  });
+
+  describe('이전 계산 결과값을 이용해 게속 계산할 수 있다.', () => {
```
👍👍👍

yujo11 commented on cypress/e2e/calculator.cy.js:
```diff
@@ -0,0 +1,222 @@
+    });
+  });
+
+  describe('맨 처음 연산자부터 클릭하면 alert창이 뜬다.', () => {
```
"맨 처음"이라는 단어를 사용하시기 보다 "숫자를 입력하지 않은 상태에서 연산자를 클릭했을 때 alert창이 뜬다."처럼 어떤 상황인지 명확하게 적어주시면 더 좋을거 같네요!

aken-you replied:
이렇게 보니 제가 쓴 테스트 문장이 구체적이지 않네요. given-when-then에 맞춰 써보도록 하겠습니다!

yujo11 commented on cypress/e2e/calculator.cy.js:
```diff
@@ -0,0 +1,222 @@
+    });
+  });
+
+  describe('사칙 연산자(+,-,x,/)를 한 번 누르고 = 버튼을 클릭하면 alert창이 뜬다.', () => {
```
이 부분도 상황이 더 명확하게 드러날 수 있도록 바꿔주시면 좋을거 같아요!

yujo11 commented on index.html:
```diff
@@ -1,5 +1,5 @@
 <!DOCTYPE html>
-<html lang="en">
+<html lang="ko">
```
👍👍👍

yujo11 commented on src/js/calculator.js:
```diff
@@ -0,0 +1,93 @@
+import { MAX_NUMBER_LENGTH, OPERATION, ALRERT_MESSAGE } from './constants.js';
+
+const calculateTotal = {
+  [OPERATION.PLUS]: (firstNumber, secondNumber) => firstNumber + secondNumber,
+  [OPERATION.MINUS]: (firstNumber, secondNumber) => firstNumber - secondNumber,
+  [OPERATION.MULTIPLY]: (firstNumber, secondNumber) => firstNumber * secondNumber,
+  [OPERATION.DIVIDE]: (firstNumber, secondNumber) => Math.floor(firstNumber / secondNumber),
+};
```
lookup table을 활용하셨군요! 👍 👍 👍

yujo11 commented on src/js/calculator.js:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  renderTotal() {
+    const { firstNumber, operation, secondNumber } = this.state;
```
자바스크립트 문법을 익숙하게 사용하시는군요! 👍 👍 👍

yujo11 commented on src/js/constants.js:
```diff
@@ -0,0 +1,15 @@
+  NOT_FIRSTNUMBER: '연산자 입력하기 전에 숫자를 입력해주세요!',
+  NOT_OVER_NUMBER_LENGTH: '3자리 이하 숫자만 입력해주세요!',
+  NOT_SECONDNUMBER: '연산자 입력한 후 숫자를 입력해주세요!',
+};
```
상수화 하신 부분이 좋네요! 👍 👍 👍

yujo11 commented on src/js/constants.js:
```diff
@@ -0,0 +1,15 @@
+export const ALRERT_MESSAGE = {
+  NOT_FIRSTNUMBER: '연산자 입력하기 전에 숫자를 입력해주세요!',
+  NOT_OVER_NUMBER_LENGTH: '3자리 이하 숫자만 입력해주세요!',
+  NOT_SECONDNUMBER: '연산자 입력한 후 숫자를 입력해주세요!',
```
소소하게 여기는 `SECOND_NUMBER`로 되어 있으면 읽기 더 좋을거 같네용

aken-you replied:
오타가 있었네요! 수정하겠습니다!

yujo11 commented on src/js/calculator.js:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  clickDigit({ dataset }) {
+    const clickedNumber = parseInt(dataset.digit, 10);
```
`parseInt`를 사용해서 형 변환을 하신 이유가 있을까용?

aken-you replied:
js에서 dataset 속성을 문자열로 가져오기 때문에 `parseInt`를 사용했습니다!

yujo11 reviewed (changes requested):
안녕하세요 수아님!

계산기 미션 리뷰를 맡게 된 조윤호입니다!

코드가 전체적으로 깔끔하고 Javascript 문법도 잘 활용하고 계셔서 코드에서 코멘트 드릴 부분이 많지 않았네요!

> calculator.js 파일 80번째 줄에 else를 썼는데 안 쓰고 할 수 있는 좋은 코드는 없을까요? return을 써서 else를 없애보려 했는데 두 경우 다 this.renderTotal() 실행해야 해서 코드가 더 늘어났습니다. 좋은 방법이 있다면 알려주세요!

함수가 하는 역할이 사칙연산자(+, -, *, /)가 들어왔을 때와 결과 연산자(=)가 들어왔을 때 이뤄지는 동작이 다르니 함수를 분리해보셔도 좋을거 같아요.

이 방법 외에는 당장 좋은 생각이 떠오르는게 없네용 ㅎㅎ...

> class method 안에서 addEventListener의 callback 함수를 사용하는 방법 중 좋은 방법은 무엇인가요? [이 사이트](https://stackoverflow.com/questions/43727516/how-adding-event-handler-inside-a-class-with-a-class-method-as-the-callback?noredirect=1&lq=1)를 통해 저는 화살표 함수로 함수를 사용했습니다.

이 부분은 일반적인 방법으로 EventListener에 Callback을 등록하면 `this` 바인딩 때문에 Callback이 등록되지 않아 해주신 질문 같은데요.

"좋은 방법"이 무엇인지는 저도 확언할 수 없네요. 첨부해주신 링크에 나오는 내용처럼 `this` 바인딩을 해주는 다양한 방법이 존재하는데요. 어떤 방법을 사용하셔도 무방할거 같습니다.

저는 개인적으로 명시적으로 `bind()`메서드를 사용해 등록하는 편입니다.

>cypress test에서 숫자, 연산자를 스스로 정하다 보니 다른 예외 경우를 확인하지 못할 수 있다고 생각했습니다. 보통 test에서 숫자나 연산자를 정하는 함수를 만드는지 궁금합니다.

이 부분은 "123 + 1 = 124"와 같은 케이스들을 직접 만들어서 쓰는지 여쭤보신거 같은데요. 

저 같은 경우는 별도의 숫자나 연산자를 따로 만들지는 않고 Edge Case라고 생각하는 부분을 최대한 고려해서 직접 만들어서 사용하고 있어요. 물론 모든 Case를 고려할 수 없기 때문에 추후 발견되는 Edge Case들은 추가하구요!

(이건 다른 얘기인데 random 값을 생성해주는 https://marketplace.visualstudio.com/items?itemName=helixquar.randomeverything 같은 VSC Extension도 종종 사용합니다)

---

리뷰와 코멘트 확인 후 다시 리뷰 요청 부탁드리겠습니다~ 고생 많으셨습니다! 👍

aken-you committed:
refactor: constants 네이밍

aken-you committed:
chore: eslint 규칙 추가

aken-you committed:
test: 테스트 리펙토링

1. given-when-then에 맞춰 테스트 문장 작성
2. 함수로 중복된 코드 정리
- calculation: 계산 결과 확인
- modifierTest: 초기화 확인
- continueCalculation: 연이어 계산이 가능한지 확인
- alertNotOverNumberLength: 글자 수 제한 경고 확인
- alertNotFirstNumber: 첫 번째 숫자 존재 확인
- alertNotSecondNumber: 두 번째 숫자 존재 확인

aken-you committed:
refactor: 글자수 제한 메소드 분리

aken-you committed:
test: Command 추가

- calculation, modifierTest, continueCalculation, alertNotOverNumberLength, alertNotFirstNumber, alertNotSecondNumber

yujo11 commented on src/js/calculator.js:
```diff
@@ -45,7 +45,7 @@ class Calculator {
 
     if (secondNumber === null) {
-      alert(ALRERT_MESSAGE.NOT_SECONDNUMBER);
+      alert(ALERT_MESSAGE.NOT_SECOND_NUMBER);
```
👍👍👍

yujo11 commented on cypress/e2e/calculator.cy.js:
```diff
@@ -5,218 +5,158 @@ describe('계산기 테스트', () => {
+    cy.clickDigit(secondNumber);
+    cy.clickOperation(OPERATION.EQUAL);
+    cy.getTotal(expectedTotal);
+  };
```
👍👍👍

yujo11 replied:
여러 개의 인자를 받는 함수의 경우 사용하는 곳에서 쉽게 사용하고, 실수를 방지하기 위해 객체로 묶어서 인자를 받도록 변경해보셔도 좋을거 같네요!

```suggestion
  const calculation = ({ firstNumber, operation, secondNumber, expectedTotal }) => {
    cy.clickDigit(firstNumber);
    cy.clickOperation(operation);
    cy.clickDigit(secondNumber);
    cy.clickOperation(OPERATION.EQUAL);
    cy.getTotal(expectedTotal);
  };
```

yujo11 commented on cypress/e2e/calculator.cy.js:
```diff
@@ -5,218 +5,158 @@ describe('계산기 테스트', () => {
-
-      cy.clickOperation(OPERATION.MULTIPLY);
-      cy.clickDigit([2]);
+  describe('계산 결과값이 나온 상태에서 사칙연산자를 클릭하고 숫자를 입력했을 때 계산이 가능하다.', () => {
```
이제는 `describe`에 적힌 설명만 봐도 어떤 내용을 테스트 하는지 명확히 알 수 있네요! 👍 👍 👍

yujo11 commented on src/js/calculator.js:
```diff
@@ -29,16 +29,21 @@ class Calculator {
     this.$operations.addEventListener('click', (e) => this.clickOperation(e.target));
   }
 
+  isOverNumberLength(newNumber) {
+    if (String(newNumber).length > MAX_NUMBER_LENGTH) {
+      alert(ALERT_MESSAGE.NOT_OVER_NUMBER_LENGTH);
+      return true;
+    }
+    return false;
+  }
```
👍👍👍

yujo11 reviewed (approved):
피드백을 빠르게 반영해주셨군요! 

계산기 미션 진행하느라 고생 많으셨습니다 수아님!

간단한 코멘트 남겼으니 확인 부탁드리겠습니다! 다음 미션도 화이팅하세요~!!! 👍 👍 👍
